File: kotlin/src/main/kotlin/com/bhengubv/circleai/faith/Faith.kt
// Faith.kt
//
// Kotlin port of CircleAI.Faith (FaithPrimitives.cs + FaithDomainContext.cs +
// FaithCompanionAdapter.cs) — the C# reference is the EXACT spec. A
// deterministic in-memory faith board: services, prayer requests, and scripture.
//
// Fidelity notes:
//   * C# `record` -> Kotlin `data class`; `DateTimeOffset` -> `Instant`.
//   * `ServicesBetween` inclusive [start, end], ASC.
//   * `RecentPrayers` newest-first, capped at `limit` (default 20).
//   * `Lookup` = first exact (tradition, book, chapter, verse) match (or null).
//   * `ByTradition` case-insensitive.

package com.bhengubv.circleai.faith

import com.bhengubv.circleai.companion.CompanionContext
import com.bhengubv.circleai.companion.CompanionProactiveEvent
import com.bhengubv.circleai.companion.CompanionTurn
import com.bhengubv.circleai.companion.ICompanionSession
import com.bhengubv.circleai.companion.InterfaceKind
import kotlinx.coroutines.flow.Flow
import java.time.Instant
import java.util.concurrent.ConcurrentHashMap

// =====================================================================
// Primitives (FaithPrimitives.cs)
// =====================================================================

/** A faith community service. Mirrors C# `FaithService`. */
data class FaithService(val serviceId: String, val communityName: String, val title: String, val startUtc: Instant, val location: String)

/** A prayer request. Mirrors C# `PrayerRequest`. */
data class PrayerRequest(val requestId: String, val author: String, val body: String, val submittedUtc: Instant, val isAnonymous: Boolean)

/** A scripture reference. Mirrors C# `ScriptureReference`. */
data class ScriptureReference(val referenceId: String, val tradition: String, val book: String, val chapter: Int, val verse: Int, val text: String)

/** Deterministic faith board. Mirrors C# `IFaithBoard`. */
interface IFaithBoard {
    fun schedule(s: FaithService)
    fun servicesBetween(start: Instant, end: Instant): List<FaithService>
    fun submitPrayer(r: PrayerRequest)
    fun recentPrayers(limit: Int = 20): List<PrayerRequest>
    fun addScripture(r: ScriptureReference)
    fun lookup(tradition: String, book: String, chapter: Int, verse: Int): ScriptureReference?
    fun byTradition(tradition: String): List<ScriptureReference>
}

/** In-memory [IFaithBoard]. Mirrors C# `InMemoryFaithBoard`. */
class InMemoryFaithBoard : IFaithBoard {
    private val services = ConcurrentHashMap<String, FaithService>()
    private val prayers = mutableListOf<PrayerRequest>()
    private val scripture = ConcurrentHashMap<String, ScriptureReference>()
    private val lock = Any()

    override fun schedule(s: FaithService) { services[s.serviceId] = s }
    override fun servicesBetween(start: Instant, end: Instant): List<FaithService> =
        services.values.filter { !it.startUtc.isBefore(start) && !it.startUtc.isAfter(end) }.sortedBy { it.startUtc }

    override fun submitPrayer(r: PrayerRequest) { synchronized(lock) { prayers.add(r) } }
    override fun recentPrayers(limit: Int): List<PrayerRequest> = synchronized(lock) {
        prayers.sortedByDescending { it.submittedUtc }.take(limit)
    }

    override fun addScripture(r: ScriptureReference) { scripture[r.referenceId] = r }
    override fun lookup(tradition: String, book: String, chapter: Int, verse: Int): ScriptureReference? =
        scripture.values.firstOrNull { it.tradition == tradition && it.book == book && it.chapter == chapter && it.verse == verse }

    override fun byTradition(tradition: String): List<ScriptureReference> =
        scripture.values.filter { it.tradition.equals(tradition, ignoreCase = true) }
}

// =====================================================================
// DomainContext (FaithDomainContext.cs)
// =====================================================================

/** Static domain context for Faith. Mirrors C# `FaithDomainContext`. */
object FaithDomainContext {
    const val SYSTEM_PROMPT_SNIPPET: String =
        "[DOMAIN: Faith] Respectful, non-denominational spiritual companion. Help with scripture study, " +
            "prayer composition, devotional content, faith community planning, and spiritual reflection " +
            "prompts. Respect all faith traditions equally. Never impose one tradition on another. Compliance: POPIA."

    val complianceFlags: List<String> = listOf("POPIA", "Non_Denominational_Respect")

    val suggestedTools: List<String> = listOf("scripture_tools", "document_editor", "calendar")
}

// =====================================================================
// CompanionAdapter (FaithCompanionAdapter.cs)
// =====================================================================

/** Wraps an [ICompanionSession] with the Faith snippet + helpers. Mirrors C# `FaithCompanionAdapter`. */
class FaithCompanionAdapter(private val inner: ICompanionSession) : ICompanionSession {
    override val sessionId: String get() = inner.sessionId
    override val identityId: String get() = inner.identityId
    override val interfaceKind: InterfaceKind get() = inner.interfaceKind
    override val history: List<CompanionTurn> get() = inner.history
    override val proactiveEvents: Flow<CompanionProactiveEvent> get() = inner.proactiveEvents

    override fun getContext(): CompanionContext = inner.getContext()
    override suspend fun refreshContextAsync() = inner.refreshContextAsync()
    override suspend fun signalFeedbackAsync(positive: Boolean, note: String?) =
        inner.signalFeedbackAsync(positive, note)
    override fun close() = inner.close()

    override suspend fun sendAsync(message: String): String = inner.sendAsync(enrich(message))
    override fun streamAsync(message: String): Flow<String> = inner.streamAsync(enrich(message))
    override suspend fun agentAsync(instruction: String): String = inner.agentAsync(enrich(instruction))

    private fun enrich(m: String): String = "${FaithDomainContext.SYSTEM_PROMPT_SNIPPET}\n\n$m"

    suspend fun generateDevotionalAsync(theme: String, tradition: String): String =
        inner.agentAsync("Write a short devotional on the theme of $theme for the $tradition tradition. Include a scripture reference, reflection, and closing prayer or meditation.")

    suspend fun studyScriptureAsync(passage: String, question: String): String =
        inner.agentAsync("Help me study $passage. Question: $question. Provide historical context, key interpretations across traditions, and practical application.")

    suspend fun composeReflectionAsync(tradition: String, occasion: String, scriptureRef: String): String =
        inner.agentAsync("Compose a 200-word reflection in the $tradition for $occasion, anchored in $scriptureRef. Warm, inclusive, devotional.")

    suspend fun draftServiceOrderAsync(tradition: String, serviceType: String, durationMinutes: Int): String =
        inner.agentAsync("Draft a $durationMinutes-minute $serviceType order of service in the $tradition. Sections, transitions, music cues, scripture readings.")

    suspend fun writePastoralCareNoteAsync(parishionerSituation: String): String =
        inner.agentAsync("Write a pastoral care note for: $parishionerSituation. Acknowledge, hold space, offer concrete next step. Avoid platitudes.")

    suspend fun findScripturePassagesAsync(tradition: String, theme: String): String =
        inner.agentAsync("Find 3 scripture passages on '$theme' in the $tradition. For each: reference, key verse text, brief context.")
}

File: swift/Sources/CircleAI/Faith.swift
// Faith.swift
//
// Port of the Faith vertical from src/CircleAI.Faith/FaithPrimitives.cs and the
// static domain-context constants from FaithDomainContext.cs:
//   • FaithService, PrayerRequest, ScriptureReference — domain records
//   • IFaithBoard                                      — services, prayers, scripture
//   • InMemoryFaithBoard                               — deterministic in-memory impl
//   • FaithDomainContext                               — system-prompt snippet + flags
//
// The Companion-facing wrapper (FaithCompanionAdapter) is an ICompanionSession
// decorator that prefixes the faith domain prompt.
//
// Porting notes:
//   • `DateTimeOffset` → `Date`.
//   • `ServicesBetween` filters StartUtc in [start, end], ordered ascending.
//   • `RecentPrayers(limit)` orders descending by SubmittedUtc, take limit (20).
//   • `Lookup` matches tradition+book+chapter+verse exactly (ordinal).
//   • `ByTradition` matches Tradition case-insensitively. All state guarded by a
//     single `NSLock`.

import Foundation

// MARK: - Records

/// A scheduled faith service.
public struct FaithService: Sendable, Equatable, Codable {
    public let serviceId: String
    public let communityName: String
    public let title: String
    public let startUtc: Date
    public let location: String

    public init(serviceId: String, communityName: String, title: String, startUtc: Date, location: String) {
        self.serviceId = serviceId
        self.communityName = communityName
        self.title = title
        self.startUtc = startUtc
        self.location = location
    }
}

/// A prayer request.
public struct PrayerRequest: Sendable, Equatable, Codable {
    public let requestId: String
    public let author: String
    public let body: String
    public let submittedUtc: Date
    public let isAnonymous: Bool

    public init(requestId: String, author: String, body: String, submittedUtc: Date, isAnonymous: Bool) {
        self.requestId = requestId
        self.author = author
        self.body = body
        self.submittedUtc = submittedUtc
        self.isAnonymous = isAnonymous
    }
}

/// A scripture reference.
public struct ScriptureReference: Sendable, Equatable, Codable {
    public let referenceId: String
    public let tradition: String
    public let book: String
    public let chapter: Int
    public let verse: Int
    public let text: String

    public init(referenceId: String, tradition: String, book: String, chapter: Int, verse: Int, text: String) {
        self.referenceId = referenceId
        self.tradition = tradition
        self.book = book
        self.chapter = chapter
        self.verse = verse
        self.text = text
    }
}

// MARK: - Contract

/// Services, prayers, and scripture references for the faith vertical.
public protocol IFaithBoard: AnyObject, Sendable {
    func schedule(_ s: FaithService)
    func servicesBetween(start: Date, end: Date) -> [FaithService]
    func submitPrayer(_ r: PrayerRequest)
    func recentPrayers(limit: Int) -> [PrayerRequest]
    func addScripture(_ r: ScriptureReference)
    func lookup(tradition: String, book: String, chapter: Int, verse: Int) -> ScriptureReference?
    func byTradition(_ tradition: String) -> [ScriptureReference]
}

public extension IFaithBoard {
    /// Convenience overload mirroring the C# default `limit = 20`.
    func recentPrayers() -> [PrayerRequest] { recentPrayers(limit: 20) }
}

// MARK: - InMemoryFaithBoard

/// Deterministic in-memory `IFaithBoard`. All state guarded by a single `NSLock`.
public final class InMemoryFaithBoard: IFaithBoard, @unchecked Sendable {
    private let lock = NSLock()
    private var services: [String: FaithService] = [:]
    private var prayers: [PrayerRequest] = []
    private var scripture: [String: ScriptureReference] = [:]

    public init() {}

    public func schedule(_ s: FaithService) {
        lock.lock(); defer { lock.unlock() }
        services[s.serviceId] = s
    }

    public func servicesBetween(start: Date, end: Date) -> [FaithService] {
        lock.lock(); defer { lock.unlock() }
        return services.values.filter { $0.startUtc >= start && $0.startUtc <= end }.sorted { $0.startUtc < $1.startUtc }
    }

    public func submitPrayer(_ r: PrayerRequest) {
        lock.lock(); defer { lock.unlock() }
        prayers.append(r)
    }

    public func recentPrayers(limit: Int = 20) -> [PrayerRequest] {
        lock.lock(); defer { lock.unlock() }
        return Array(prayers.sorted { $0.submittedUtc > $1.submittedUtc }.prefix(limit))
    }

    public func addScripture(_ r: ScriptureReference) {
        lock.lock(); defer { lock.unlock() }
        scripture[r.referenceId] = r
    }

    public func lookup(tradition: String, book: String, chapter: Int, verse: Int) -> ScriptureReference? {
        lock.lock(); defer { lock.unlock() }
        return scripture.values.first { $0.tradition == tradition && $0.book == book && $0.chapter == chapter && $0.verse == verse }
    }

    public func byTradition(_ tradition: String) -> [ScriptureReference] {
        lock.lock(); defer { lock.unlock() }
        return scripture.values.filter { $0.tradition.caseInsensitiveCompare(tradition) == .orderedSame }
    }
}

// MARK: - FaithDomainContext

/// Static domain-context constants for the faith vertical.
public enum FaithDomainContext {
    public static let systemPromptSnippet = "[DOMAIN: Faith] Respectful, non-denominational spiritual companion. Help with scripture study, prayer composition, devotional content, faith community planning, and spiritual reflection prompts. Respect all faith traditions equally. Never impose one tradition on another. Compliance: POPIA."
    public static let complianceFlags: [String] = ["POPIA", "Non_Denominational_Respect"]
    public static let suggestedTools: [String] = ["scripture_tools", "document_editor", "calendar"]
}

// MARK: - FaithCompanionAdapter

/// An `ICompanionSession` decorator that prepends the faith domain system prompt
/// to every conversational call and adds faith helper methods.
/// Port of `CircleAI.Faith.FaithCompanionAdapter`. Identity/context/feedback are
/// forwarded to the inner session; proactive events forward through the inner
/// session's `proactiveEvents` stream (the Swift protocol has no disposal).
public final class FaithCompanionAdapter: ICompanionSession, @unchecked Sendable {
    private let inner: ICompanionSession

    public init(_ inner: ICompanionSession) {
        self.inner = inner
    }

    public var sessionId: String { inner.sessionId }
    public var identityId: String { inner.identityId }
    public var interface: InterfaceKind { inner.interface }
    public var history: [CompanionTurn] { inner.history }

    public func getContext() -> CompanionContext { inner.getContext() }
    public func refreshContext() async throws { try await inner.refreshContext() }
    public func signalFeedback(positive: Bool, note: String?) async throws {
        try await inner.signalFeedback(positive: positive, note: note)
    }
    public var proactiveEvents: AsyncStream<CompanionProactiveEvent> { inner.proactiveEvents }

    public func send(_ message: String) async throws -> String { try await inner.send(enrich(message)) }
    public func stream(_ message: String) -> AsyncStream<String> { inner.stream(enrich(message)) }
    public func agent(_ instruction: String) async throws -> String { try await inner.agent(enrich(instruction)) }

    private func enrich(_ m: String) -> String { "\(FaithDomainContext.systemPromptSnippet)\n\n\(m)" }

    // ── Faith helpers ─────────────────────────────────────────────────────────

    /// Generate a devotional (C# `GenerateDevotionalAsync`).
    public func generateDevotional(theme: String, tradition: String) async throws -> String {
        try await inner.agent(
            "Write a short devotional on the theme of \(theme) for the \(tradition) tradition. Include a scripture reference, reflection, and closing prayer or meditation.")
    }

    /// Study a scripture passage (C# `StudyScriptureAsync`).
    public func studyScripture(passage: String, question: String) async throws -> String {
        try await inner.agent(
            "Help me study \(passage). Question: \(question). Provide historical context, key interpretations across traditions, and practical application.")
    }

    /// Compose a reflection (C# `ComposeReflectionAsync`).
    public func composeReflection(tradition: String, occasion: String, scriptureRef: String) async throws -> String {
        try await inner.agent(
            "Compose a 200-word reflection in the \(tradition) for \(occasion), anchored in \(scriptureRef). Warm, inclusive, devotional.")
    }

    /// Draft an order of service (C# `DraftServiceOrderAsync`).
    public func draftServiceOrder(tradition: String, serviceType: String, durationMinutes: Int) async throws -> String {
        try await inner.agent(
            "Draft a \(durationMinutes)-minute \(serviceType) order of service in the \(tradition). Sections, transitions, music cues, scripture readings.")
    }

    /// Write a pastoral care note (C# `WritePastoralCareNoteAsync`).
    public func writePastoralCareNote(parishionerSituation: String) async throws -> String {
        try await inner.agent(
            "Write a pastoral care note for: \(parishionerSituation). Acknowledge, hold space, offer concrete next step. Avoid platitudes.")
    }

    /// Find scripture passages on a theme (C# `FindScripturePassagesAsync`).
    public func findScripturePassages(tradition: String, theme: String) async throws -> String {
        try await inner.agent(
            "Find 3 scripture passages on '\(theme)' in the \(tradition). For each: reference, key verse text, brief context.")
    }
}
